File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/homelist/HomeTrendingCategoryListViewModel.java
package com.dewonderstruck.apps.ashx0.viewmodel.homelist;

import com.dewonderstruck.apps.ashx0.repository.category.CategoryRepository;
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData;
import com.dewonderstruck.apps.ashx0.utils.Utils;
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel;
import com.dewonderstruck.apps.ashx0.viewobject.Category;
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource;
import com.dewonderstruck.apps.ashx0.viewobject.holder.CategoryParameterHolder;
import com.dewonderstruck.apps.ashx0.viewobject.holder.ProductParameterHolder;

import java.util.List;

import javax.inject.Inject;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

public class HomeTrendingCategoryListViewModel extends PSViewModel {

    private final LiveData<Resource<List<Category>>> homeTrendingCategoryListData;
    private MutableLiveData<HomeTrendingCategoryListViewModel.TmpDataHolder> homeTrendingCategoryListobj = new MutableLiveData<HomeTrendingCategoryListViewModel.TmpDataHolder>();

    private final LiveData<Resource<Boolean>> homeTrendingCategoryLoadNetworkData;
    private final MutableLiveData<HomeTrendingCategoryListViewModel.TmpDataHolder> homeTrendingCategoryLoadNetworkListObj = new MutableLiveData<HomeTrendingCategoryListViewModel.TmpDataHolder>();

    public ProductParameterHolder productParameterHolder = new ProductParameterHolder();

    public CategoryParameterHolder categoryParameterHolder = new CategoryParameterHolder().getTrendingCategories();

    @Inject
    HomeTrendingCategoryListViewModel(CategoryRepository repository) {
        Utils.psLog("Inside ProductViewModel");

        homeTrendingCategoryListData = Transformations.switchMap(homeTrendingCategoryListobj, obj -> {
            if (obj == null) {
                return AbsentLiveData.create();
            }

            return repository.getAllSearchCategory(obj.categoryParameterHolder, obj.limit, obj.offset);
        });

        homeTrendingCategoryLoadNetworkData = Transformations.switchMap(homeTrendingCategoryLoadNetworkListObj, obj -> {
            if (obj == null) {
                return AbsentLiveData.create();
            }

            return repository.getNextSearchCategory(obj.limit, obj.offset, obj.categoryParameterHolder);
        });
    }

    public void setHomeTrendingCatrgoryListDataObj(CategoryParameterHolder categoryParameterHolder, String limit, String offset) {

        HomeTrendingCategoryListViewModel.TmpDataHolder tmpDataHolder = new HomeTrendingCategoryListViewModel.TmpDataHolder();
        tmpDataHolder.limit = limit;
        tmpDataHolder.offset = offset;
        tmpDataHolder.categoryParameterHolder = categoryParameterHolder;
        homeTrendingCategoryListobj.setValue(tmpDataHolder);

    }

    public LiveData<Resource<List<Category>>> getHomeTrendingCategoryListData() {
        return homeTrendingCategoryListData;
    }


    public void setHomeTrendingCategoryLoadNetworkObj(String loginUserId, String limit, String offset, CategoryParameterHolder categoryParameterHolder) {
        if (!isLoading) {
            HomeTrendingCategoryListViewModel.TmpDataHolder tmpDataHolder = new HomeTrendingCategoryListViewModel.TmpDataHolder();
            tmpDataHolder.loginUserId = loginUserId;
            tmpDataHolder.limit = limit;
            tmpDataHolder.offset = offset;
            tmpDataHolder.categoryParameterHolder = categoryParameterHolder;
            homeTrendingCategoryLoadNetworkListObj.setValue(tmpDataHolder);

            // start loading
            setLoadingState(true);
        }
    }

    public LiveData<Resource<Boolean>> getHomeTrendingCategoryLoadNetworkData() {
        return homeTrendingCategoryLoadNetworkData;
    }

    class TmpDataHolder {
        public String loginUserId = "";
        public String limit = "";
        public String offset = "";
        public String shopId = "";
        public String orderBy = "";
        public CategoryParameterHolder categoryParameterHolder;

    }
}

File: app/src/main/java/com/dewonderstruck/apps/ashx0/viewmodel/homelist/HomeTrendingCategoryListViewModel.kt
package com.dewonderstruck.apps.ashx0.viewmodel.homelist

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import com.dewonderstruck.apps.ashx0.repository.category.CategoryRepository
import com.dewonderstruck.apps.ashx0.utils.AbsentLiveData
import com.dewonderstruck.apps.ashx0.utils.Utils
import com.dewonderstruck.apps.ashx0.viewmodel.common.PSViewModel
import com.dewonderstruck.apps.ashx0.viewobject.Category
import com.dewonderstruck.apps.ashx0.viewobject.common.Resource
import com.dewonderstruck.apps.ashx0.viewobject.holder.CategoryParameterHolder
import com.dewonderstruck.apps.ashx0.viewobject.holder.ProductParameterHolder
import javax.inject.Inject

class HomeTrendingCategoryListViewModel @Inject internal constructor(repository: CategoryRepository) : PSViewModel() {
    val homeTrendingCategoryListData: LiveData<Resource<List<Category>>>
    private val homeTrendingCategoryListobj = MutableLiveData<TmpDataHolder>()
    val homeTrendingCategoryLoadNetworkData: LiveData<Resource<Boolean>>
    private val homeTrendingCategoryLoadNetworkListObj = MutableLiveData<TmpDataHolder>()
    @JvmField
    var productParameterHolder = ProductParameterHolder()
    @JvmField
    var categoryParameterHolder = CategoryParameterHolder().trendingCategories
    fun setHomeTrendingCatrgoryListDataObj(categoryParameterHolder: CategoryParameterHolder?, limit: String, offset: String) {
        val tmpDataHolder = TmpDataHolder()
        tmpDataHolder.limit = limit
        tmpDataHolder.offset = offset
        tmpDataHolder.categoryParameterHolder = categoryParameterHolder
        homeTrendingCategoryListobj.value = tmpDataHolder
    }

    fun setHomeTrendingCategoryLoadNetworkObj(loginUserId: String, limit: String, offset: String, categoryParameterHolder: CategoryParameterHolder?) {
        if (!isLoading) {
            val tmpDataHolder = TmpDataHolder()
            tmpDataHolder.loginUserId = loginUserId
            tmpDataHolder.limit = limit
            tmpDataHolder.offset = offset
            tmpDataHolder.categoryParameterHolder = categoryParameterHolder
            homeTrendingCategoryLoadNetworkListObj.value = tmpDataHolder

            // start loading
            setLoadingState(true)
        }
    }

    internal inner class TmpDataHolder {
        var loginUserId = ""
        var limit = ""
        var offset = ""
        var shopId = ""
        var orderBy = ""
        var categoryParameterHolder: CategoryParameterHolder? = null
    }

    init {
        Utils.psLog("Inside ProductViewModel")
        homeTrendingCategoryListData = Transformations.switchMap(homeTrendingCategoryListobj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<List<Category>>>()
            }
            repository.getAllSearchCategory(obj.categoryParameterHolder, obj.limit, obj.offset)
        }
        homeTrendingCategoryLoadNetworkData = Transformations.switchMap(homeTrendingCategoryLoadNetworkListObj) { obj: TmpDataHolder? ->
            if (obj == null) {
                return@switchMap AbsentLiveData.create<Resource<Boolean>>()
            }
            repository.getNextSearchCategory(obj.limit, obj.offset, obj.categoryParameterHolder)
        }
    }
}